zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1184,13 +1192,21 @@ public void run() {
                         to = connectTimeout - clientCnxnSocket.getIdleRecv();
                     }
 
-                    if (to <= 0) {
+                    int expiration = expirationTimeout - clientCnxnSocket.getIdleRecv();
+                    if (expiration <= 0) {
                         String warnInfo = String.format(
                             "Client session timed out, have not heard from server in %dms for session id 0x%s",
                             clientCnxnSocket.getIdleRecv(),
                             Long.toHexString(sessionId));
                         LOG.warn(warnInfo);
+                        changeZkState(States.CLOSED);
```

Comment:
Why have you added this here?

Reply:
`States.CLOSED` is the terminal state for `KeeperState.Expired`. We do this for server side `KeeperState.Expired`, so does client side `KeeperState.Expired`. This way we won't get extra `KeeperState.Disconnected` event, which fired already in `cleanAndNotifyState`, before `KeeperState.Expired`.

https://github.com/apache/zookeeper/blob/bc9afbf8ef1bc6156643d3d05c87fcf8411e9d8f/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java#L1359-L1370